src/main/java/com/redhat/jenkins/plugins/ci/CIBuildTrigger.java

```diff
@@ -175,63 +180,75 @@ private void startTriggerThread() {
 			}
 		}
 		try {
-			stopTriggerThread();
-			JMSMessagingProvider provider = GlobalCIConfiguration.get()
-					.getProvider(providerName);
-			if (provider == null) {
-				log.log(Level.SEVERE, "Failed to locate JMSMessagingProvider with name "
-						+ providerName + ". You must update the job configuration. Trigger not started.");
-				return;
+			synchronized(getLock(job.getFullName())) {
+				if (stopTriggerThread(job.getFullName()) == null) {
+					JMSMessagingProvider provider = GlobalCIConfiguration.get()
+							.getProvider(providerName);
+					if (provider == null) {
+						log.log(Level.SEVERE, "Failed to locate JMSMessagingProvider with name "
+								+ providerName + ". You must update the job configuration. Trigger not started.");
+						return;
+					}
+					CITriggerThread trigger = new CITriggerThread(provider, overrides, job
+							.getFullName(), selector, getChecks());
+					trigger.setName("CIBuildTrigger-" + job.getFullName() + "-" + provider.getClass().getSimpleName());
+					trigger.setDaemon(true);
+					trigger.start();
+					log.info("Adding thread: " + trigger.getId());
+					triggerInfo.put(job.getFullName(), trigger);
+				}
 			}
-			CITriggerThread trigger = new CITriggerThread(provider, overrides, job
-					.getFullName(), selector, getChecks());
-			trigger.setName("CIBuildTrigger-" + job.getFullName() + "-" + provider.getClass().getSimpleName());
-			trigger.setDaemon(true);
-			trigger.start();
-			log.fine("Adding thread: " + trigger.getId());
-			triggerInfo.put(job.getFullName(), trigger);
 		} catch (Exception e) {
 			log.log(Level.SEVERE, "Unhandled exception in trigger start.", e);
 		}
 	}
 
-    private void stopTriggerThread() {
-        if (job != null) {
-            stopTriggerThread(job.getFullName());
-        }
-    }
+    private CITriggerThread stopTriggerThread(String fullName) {
+		synchronized(getLock(fullName)) {
```

Comment:
I would add a comment here indicating that the thread is created *only* for comparison purposes, and it is never started. This confused me at first.